### PR TITLE
Constrain Study ID by regex

### LINKED ISSUE #72: Constrain Study ID by regex
The study ID can currently be configured in the active admin, but only a single study ID is allowed. The administrator should be able to specify a range of study IDs, using a regular expression.

### PR DESCRIPTION
**Pull request by DoxasticFox**
Closes #72.

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -34,6 +34,7 @@ gem 'wicked_pdf'
 gem 'wkhtmltopdf-binary'
 gem 'activeadmin_addons'
 gem 'bootstrap-datepicker-rails'
+gem 'regexp-examples', '~> 1.5', '>= 1.5.1'
```
Comment (DoxasticFox) [inline Gemfile:37]: This gem converts a regular expression to some strings which match that regular expression. I thought it'd be helpful to show the administrator what their regex does.

Screenshot_20221110_054034

File: db/data.yml [hunk 2]
```diff
@@ -1,5 +1,5 @@
 - StudyCode:
-  - title: A1543457
+  - title: "\\AA[0-4]{1}[0-9]{1}[2-4]{1}[0-9]{4}\\z"
```
Comment (DoxasticFox) [inline db/data.yml:2]: Note that I moved this from `app/models/study_code.rb`.

File: app/lib/user_date_validator.rb [hunk 3]
```diff
@@ -13,7 +13,12 @@ def child_date_of_birth_in_future
 
   def check_study_code
     codes = StudyCode.pluck(:title)
-    codes.include?(study_id)? true : errors.add(:study_id, "Study ID not present in our system. Please try with different one.") && false
+    if codes.all? { |code| Regexp.new(code).match(study_id) }
```
Comment (DoxasticFox) [inline app/lib/user_date_validator.rb:16]: The validation on `StudyCode` is such that it should contain exactly one record. So, arguably, the `if` statement's condition should be something like `Regexp.new(codes[0]).match(study_id)` (with an exception raised when `codes.length != 1`.

File: app/admin/study_code.rb [hunk 4]
```diff
@@ -1,33 +1,32 @@
...
+  filter :title
 
-    form do |f|
-        f.inputs do
-            f.input :title
-        end
-        f.actions
+  form do |f|
+    f.inputs do
+      f.input :title, hint: hint(f.object.title)
```
Comment (DoxasticFox) [inline app/admin/study_code.rb:28]: With the exception of this line, all the changes in this file are just formatting.

File: Gemfile (hunk 1 above)
Comment (ammazzaw) [reply Gemfile:37]: Neat! Can it be used as generator for tests?

Comment (DoxasticFox) [reply Gemfile:37]: Property-based tests are cool but I don't see a way to use them here. Do you have a concrete idea for how I could test?

File: spec/models/study_code_spec.rb [hunk 5]
```diff
@@ -2,44 +2,11 @@
...
-      let(:study_code) { StudyCode.create(title: 'B1523456') }
+    context 'when StudyCode is not a valid regex' do
+      let(:study_code) { StudyCode.create(title: 'B1523456[') }
 
       it { expect(study_code).to be_a(StudyCode) }
-      it { expect(study_code.errors).not_to be_empty }
-      it { expect(study_code.errors[:title]).to eq(['Invalid study code format']) }
-    end
-
-    context 'when first number is not between 0 to 4 after capital A' do
```
Comment (ammazzaw) [inline spec/models/study_code_spec.rb:13]: I can't believe no one put a regex here earlier 😹

**Review by ammazzaw — APPROVED** [review]

**Pull request merged**